docs/proposed/psa-driver-interface.md

```diff
@@ -345,6 +390,99 @@ Transparent drivers may provide the following key management entry points:
 
 Transparent drivers are not involved when importing, exporting, copying or destroying keys, or when generating or deriving symmetric keys.
 
+### Random generation entry points
+
+A transparent driver may provide an operation family that allows it to generate random data. The random generation mechanism must obey the following requirements:
+
+* The random output must be of cryptographic quality, with a uniform distribution. Therefore, if the random generator includes an entropy source, this entropy source must be fed through a CSPRNG (cryptographically secure pseudo-random number generator).
+* Random generation is expected to be fast. (If a device can provide entropy but is slow at generating random data, declare it as an [entropy driver](#entropy-collection-entry-point) instead.)
+* The random generator must be able to incorporate entropy provided by an outside source.
+* The random generator may either be deterministic (in the sense that it always returns the same data when given the same entropy inputs) or non-deterministic (including its own entropy source). In other words, this interface is suitable both for PRNG (pseudo-random number generator, also known as DRBG (deterministic random bit generator)) and for NRBG (non-deterministic random bit generator).
+
+If no driver implements the random generation entry point family, the core provides an unspecified random generation mechanism.
+
+This operation family requires the following type, entry points and parameters (TODO: where exactly are the parameters in the JSON structure?):
+
+* Type `"random_context_t"`: the type of a random generation context.
+* `"init_random"` (entry point, optional): if this function is present, [the core calls it once](#random-generator-initialization) after allocating a `"random_context_t"` object.
+* `"add_entropy"` (entry point, optional): the core calls this function to [inject entropy](#entropy-injection). This entry point is optional if the driver is for a peripheral that includes an entropy source of its own, however [random generator drivers without entropy injection](#random-generator-drivers-without-entropy-injection) have limited portability since they can only be used on platforms with no other entropy source.
+* `"get_random"` (entry point, mandatory): the core calls this function whenever it needs to [obtain random data](#the-get_random-entry-point).
+* `"initial_entropy_size"` (integer, optional): the minimum number of bytes of entropy that the core must supply before the driver can output random data. This can be `0` if the driver is for a peripheral that includes an entropy source of its own. If omitted, the value is `0`.
```

Comment:
This value comes from the driver. If the device includes an entropy source, it doesn't need external entropy, so it should declare `initial_entropy_size` = 0. It still makes sense to have a nonzero `reseed_entropy_size` to convey “if you're going to reseed anyway, this is how many bytes I expect”.

If there are other entropy sources on the device, the core would call `add_entropy` initially, even if `initial_entropy_size` is 0.

Reply:
> If the device includes an entropy source, it doesn't need external entropy, so it should declare `initial_entropy_size` = 0

Yes, I have no problem with the driver being able to declare that explicitly, I'm more concerned about making it the default. 

> If there are other entropy sources on the device, the core would call `add_entropy` initially, even if `initial_entropy_size` is 0.

Ok, so apparently I don't understand what `initial_entropy_size == 0` means. I thought it meant "dear core, please don't bother feeding me entropy before calling `get_random`". Now you're telling me the core will call `add_entropy` anyway - how many times will it call it? up to a total of how many bytes of entropy?

Reply:
It's not exactly “don't bother feeding me entropy”, but rather “you don't need to feed me entropy”.

Maybe it would be better to have a separate boolean parameter `needs_initial_entropy`? But I think it would be redundant with `initial_entropy_size` being 0.

Reply:
Ok, thanks for clarifying the meaning of `initial_entropy_size == 0`. I don't think an extra parameter is required. My main question is really about the default value, and I don't think you've answered that so far.

Defaulting to `0` means "a driver doesn't need entropy unless it explicitly declared it needed it". It would seem much safer to me to take the approach that "a driver needs entropy unless it explicitly declared it didn't".

By defaulting to "doesn't need entropy" we run the risk of silently getting non-random bytes if the driver forgot to declare an `initial_entropy_size`. If we defaulted to "the expected security strength of the device" I don't see what serious issue it would cause: if the driver forgets to declare it doesn't need entropy and there are no other entropy sources, then we would get an error, which would be easily fixed by declaring an `initial_entropy_size` of `0`.

Reply:
Even if `initial_entropy_size` is 0, the core is allowed to make an initial call `add_entropy`. The “security strength of the device” is a policy choice, not something that this specification mandates. So defaulting to “the expected security strength of the device” would mean defaulting to a value chosen by the core anyway.

Reply:
I understand that the core is allowed to make an initial call to `add_entropy`, with an amount of entropy of its choosing. The question is what happens when it can't, because it has no entropy sources. I think we can agree that in this case, either the driver has its own source of entropy, or we'd like to give an error (rather than output predictable "random" bytes).

My point is that we shouldn't just assume that a driver has its own source of entropy: we should require the driver to declare it explicitly, and I think *omitting* a parameter is much less explicit than setting it to `0`.

Or maybe we can turn the question: why do you think `0` is better than other choices for the default value?

Reply:
Fair enough, then I guess the parameter should be mandatory (unless `add_entropy` is omitted).

Reply:
Can we make it always mandatory and it must be `0` is `add_entropy` is omitted? That would provide an explicit confirmation that `add_entropy` was omitted on purpose rather than by accident.